### PR TITLE
text: implements call instruction

### PR DESCRIPTION
**Pull request by codefromthecrypt**
getting closer to being able to remove checked-in wasm binaries

### CODE REVIEW

File: wasi/wasi.go [hunk 1]
```diff
@@ -358,8 +358,8 @@ func (w *WASIEnvironment) args_sizes_get(ctx *wasm.HostFunctionCallContext, args
 //     The total number of bytes written there is the value that args_sizes_get returns in argsBufSizePtr. The caller must ensure that
 //     the buffer has the enough size.
 //     Each *C.char pointer that can be obtained from argsPtr points to the beginning of each of these null-terminated strings.
-// See: https://github.com/WebAssembly/WASI/blob/snapshot-01/phases/snapshot/docs.md#-args_getargv-pointerpointeru8-argv_buf-pointeru8---errno
-// See: https://en.wikipedia.org/wiki/Null-terminated_string
```
Comment (codefromthecrypt) [inline wasi/wasi.go:362]: I noticed I suggested something not conventional before as we always didn't use a colon!

File: wasm/text/decoder.go [hunk 2]
```diff
@@ -585,11 +585,17 @@ func (p *moduleParser) resolveFunctionIndices(module *wasm.Module) error {
 		if err != nil {
 			return err
 		}
+		importCount := p.funcNamespace.count - uint32(len(module.CodeSection))
 		switch unresolved.section {
-		case wasm.SectionIDStart:
-			module.StartSection = &target
+		case wasm.SectionIDCode:
+			if target > 255 {
+				return errors.New("TODO: unresolved function indexes that don't fit in a byte")
```
Comment (codefromthecrypt) [inline wasm/text/decoder.go:592]: to keep PR small

File: wasm/text/func_parser.go [hunk 3]
```diff
@@ -111,31 +115,42 @@ func (p *funcParser) end() (tokenParser, error) {
 	return p.onFunc(p.currentTypeIdx, code, p.currentParamNames)
 }
 
+// parseFuncIndex parses an index in the function namespace and appends it to the currentBody. If it was an ID, a
+// placeholder byte(0) is added instead and will be resolved later.
+func (p *funcParser) parseFuncIndex(tok tokenType, tokenBytes []byte, line, col uint32) (tokenParser, error) {
+	funcIdx := p.funcNamespace.count // TODO: move all function logic here and use the code index, not including imports
```
Comment (codefromthecrypt) [inline wasm/text/func_parser.go:121]: to keep PR small

**Review by codefromthecrypt — COMMENTED** [review]
I will address these later

File: wasm/text/decoder.go [hunk 4]
```diff
@@ -585,11 +585,17 @@ func (p *moduleParser) resolveFunctionIndices(module *wasm.Module) error {
 		if err != nil {
 			return err
 		}
+		importCount := p.funcNamespace.count - uint32(len(module.CodeSection))
```
Comment (mathetake) [inline wasm/text/decoder.go:588]: can this be calculated outside of loop?

**Review by mathetake — APPROVED** [review]
This is exciting!

File: wasm/text/decoder.go (hunk 4 above)
Comment (codefromthecrypt) [reply wasm/text/decoder.go:588]: indeed. the other TODO, when fixed, will fix this!

**Comment by codefromthecrypt** [comment]
thanks for the look @mathetake !

**Pull request merged**